File: Example/model/GymChain.java
package model;

import java.util.ArrayList;
import java.util.List;

import model.gym.Gym;
import model.members.Member;
import model.members.MembershipLevel;

public class GymChain {

    /**
     * Gyms of this gym chain.
     */
    private List<Gym> gyms;
    /**
     * Members of this gym chain.
     */
    private List<Member> members;

    /**
     * Default constructor.
     */
    public GymChain() {
        gyms = new ArrayList<>();
        members = new ArrayList<>();
    }

    /**
     * Accessor of gyms.
     * @return gyms
     */
    public List<Gym> getGyms() {
        return gyms;
    }

    /**
     * Finds the gym according to the given location name.
     * @param location given location name
     * @return gym if found, null if not found
     */
    public Gym findGym(String location) {
        for (Gym gym : gyms) {
            if (gym.getLocation().toLowerCase().equals(location.toLowerCase())) {
                return gym;
            }
        }
        // Not found
        return null;
    }

    /**
     * Accessor of members.
     * @return members
     */
    public List<Member> getMembers() {
        return members;
    }

    /**
     * Finds the member according to the given name.
     * @param name given name
     * @return member if found, null if not found
     */
    public Member findMember(String name) {
        for (Member member : members) {
            if (member.getName().toLowerCase().equals(name.toLowerCase())) {
                return member;
            }
        }
        // Not found
        return null;
    }

    /**
     * Determines whether a member can enter a gym.
     * @param memberName name of the member
     * @param gymLocation location name of the gym
     * @return whether the member can enter the gym
     */
    public boolean canMemberEnterGym(String memberName, String gymLocation) {
        Member member = findMember(memberName);
        Gym gym = findGym(gymLocation);
        if ((member == null) || (gym == null)) {
            return false;
        }

        if (member.getLevel() == MembershipLevel.DELUXE) {
            return true;
        }
        return !gym.getClass().getSimpleName().equals("DeluxeGym");
    }

    /**
     * Adds the given gym to this gym chain.
     * @param gym gym to add
     */
    public void addGym(Gym gym) {
        if (gym == null) {
            throw new IllegalArgumentException("The gym should not be null.");
        }
        gyms.add(gym);
    }

    /**
     * Signs up the given member to this gym chain.
     * @param member member to add
     */
    public void signUpMember(Member member) {
        if (member == null) {
            throw new IllegalArgumentException("The member should not be null.");
        }
        members.add(member);
    }

}

File: 3-UML Diagrams/Gym Example/Python/gym_chain.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Gym-chain module.
"""

from typing import List

from model.gym import DeluxeGym, Gym
from model.members import Member, MembershipLevel


class GymChain:
    """
    GymChain class.
    """
    __slots__ = ['_gyms', '_members']

    def __init__(self):
        """
        Default constructor.
        """
        self._gyms = []
        self._members = []

    @property
    def gyms(self) -> List[Gym]:
        return self._gyms

    def find_gym(self, location: str) -> Gym:
        """
        Finds the gym according to the given location.
        :param location: str
        :return: Gym
        """
        for gym in self._gyms:
            if gym.location.lower() == location.lower():
                return gym
        # Not found
        return None

    @property
    def members(self) -> List[Member]:
        return self._members

    def find_member(self, name: str) -> Member:
        """
        Finds the member according to the given member.
        :param name: str
        :return: Member
        """
        for member in self._members:
            if member.lower() == name.lower():
                return member
        # Not found
        return None

    def can_member_enter_gym(self, member_name, gym_location):
        member = self.find_member(member_name)
        gym = self.find_gym(gym_location)
        if not member or not gym:
            return False

        if member.membership_level == MembershipLevel.DELUXE:
            return True
        return type(gym).__name__ != DeluxeGym.__name__

    def add_gym(self, gym: Gym) -> None:
        """
        Adds the given gym to this gym chain.
        :param gym: Gym
        :return: None
        """
        if not gym:
            print('The gym should not be null.')
            return
        self._gyms.append(gym)

    def sign_up_member(self, member: Member) -> None:
        """
        Signs up the given member to this gym chain.
        :param member: Member
        :return: None
        """
        if not member:
            print('The member should not be null.')
            return
        self._members.append(member)
